ruff_lsp/server.py

```diff
@@ -178,6 +207,149 @@ class VersionModified(NamedTuple):
+            path=path,
+            kind=DocumentKind.Notebook,
+            source=_create_notebook_json(notebook_document),
+            version=notebook_document.version,
+        )
+
+    @classmethod
+    def from_uri(cls, uri: str) -> Self:
+        """Create a `Document` from the given URI.
+
+        The URI can be a file URI, a notebook URI, or a cell URI. The function will
+        try to get the notebook document first, and if that fails, it will fallback
+        to the text document.
+        """
+        notebook_document = LSP_SERVER.workspace.get_notebook_document(cell_uri=uri)
+        if notebook_document is None:
+            notebook_document = LSP_SERVER.workspace.get_notebook_document(
+                notebook_uri=uri
+            )
+        if notebook_document is None:
+            text_document = LSP_SERVER.workspace.get_text_document(uri)
+            return cls.from_text_document(text_document)
+        else:
+            return cls.from_notebook_document(notebook_document)
+
+    def is_stdlib_file(self) -> bool:
+        """Return True if the document belongs to standard library."""
+        return utils.is_stdlib_file(self.path)
+
+    def is_notebook_file(self) -> bool:
+        """Return True if the document belongs to a Notebook or a cell in a Notebook."""
+        return self.kind is DocumentKind.Notebook or self.path.endswith(".ipynb")
+
+
+SourceValue = str | list[str]
+
+
+class CodeCell(TypedDict):
+    """A code cell in a Jupyter notebook."""
+
+    cell_type: Literal["code"]
+    metadata: Any
+    outputs: list[Any]
+    source: SourceValue
+
+
+class MarkdownCell(TypedDict):
+    """A markdown cell in a Jupyter notebook."""
+
+    cell_type: Literal["markdown"]
+    metadata: Any
+    source: SourceValue
+
+
+class Notebook(TypedDict):
+    """The JSON representation of a Notebook Document."""
+
+    metadata: Any
+    nbformat: int
+    nbformat_minor: int
+    cells: list[CodeCell | MarkdownCell]
+
+
+def _create_notebook_json(notebook_document: NotebookDocument) -> str:
+    """Create a JSON representation of the given Notebook Document."""
+    cells: list[CodeCell | MarkdownCell] = []
+    for notebook_cell in notebook_document.cells:
+        cell_document = LSP_SERVER.workspace.get_text_document(notebook_cell.document)
+        if notebook_cell.kind is NotebookCellKind.Code:
+            cells.append(
+                {
+                    "cell_type": "code",
+                    "metadata": None,
+                    "outputs": [],
+                    "source": cell_document.source,
+                }
+            )
+        else:
+            cells.append(
+                {
+                    "cell_type": "markdown",
+                    "metadata": None,
+                    "source": cell_document.source,
+                }
+            )
+    return json.dumps(
+        {
+            "metadata": {},
+            "nbformat": 4,
+            "nbformat_minor": 5,
```

Comment:
The version doesn't matter in this context as Ruff mainly uses it to add the cell ID to maintain consistency when Ruff writes the notebook content back on disk. For VSCode, we aren't going to touch the actual file content.